presto-main/src/main/java/com/facebook/presto/cost/UnnestStatsRule.java

```diff
@@ -0,0 +1,75 @@
+
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.ComposableStatsCalculator.Rule;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.UnnestNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.unnest;
+import static java.lang.Double.NaN;
+
+public class UnnestStatsRule
+        implements Rule<UnnestNode>
+{
+    private static final int UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION = 1;
+    @Override
+    public Pattern<UnnestNode> getPattern()
+    {
+        return unnest();
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(UnnestNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        // Since we don't have stats for cardinality about the unnest column, we cannot estimate the row count.
+        // However, when the source row count is low, the error would not matter much in query optimization.
+        // Thus we'd still populate the inaccurate numbers just so stats are populated to enable optimization
+        // potential.
+        PlanNodeStatsEstimate.Builder calculatedStats = PlanNodeStatsEstimate.builder();
+        if (sourceStats.getOutputRowCount() <= UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION) {
+            calculatedStats.setOutputRowCount(sourceStats.getOutputRowCount());
+        }
+        else {
+            calculatedStats.setOutputRowCount(NaN);
+        }
+
+        for (Symbol symbol : node.getReplicateSymbols()) {
+            calculatedStats.addSymbolStatistics(symbol, sourceStats.getSymbolStatistics(symbol));
+        }
+        for (Map.Entry<Symbol, List<Symbol>> entry : node.getUnnestSymbols().entrySet()) {
+            List<Symbol> unnestToSymbols = entry.getValue();
+            SymbolStatsEstimate stats = sourceStats.getSymbolStatistics(entry.getKey());
+            for (Symbol symbol : unnestToSymbols) {
+                // This is a very conservative way on estimating stats after unnest. We assume each symbol
+                // after unnest would have as much data as the symbol before unnest. This would over
+                // estimate, which are more likely to mean we'd loose an optimization opportunity, but at
+                // least it won't cause false optimizations.
+                calculatedStats.addSymbolStatistics(
+                        symbol,
+                        SymbolStatsEstimate.builder()
+                                .setNullsFraction(stats.getNullsFraction())
+                                .setAverageRowSize(stats.getAverageRowSize())
+                                .setDistinctValuesCount(stats.getDistinctValuesCount())
```

Comment:
In fact there could be way more distinct values than in the input column. Consider column values like `ARRAY[1,2,3]` and `ARRAY[4,5,6]`. There are 2 distinct arrays, but when un-nested there are 6 distinct values.